### DISCUSSION TITLE
New Feature: Verified GT students can add reviews!

### DISCUSSION THREAD

**Discussion by m4ttsch** [oms-tech/reviews#39] [Announcements]
## Adding Course Reviews ✏️ 

Since taking over ownership of the OMSCentral project, I've been working hard to recreate the core functionality so you can continue to use this powerful tool.

As of today, I've released one of the most important features: **Adding reviews**!

The add review form asks for the same data that you may have seen in previous iterations of the site:
- Course selection
- Semester selection
- 1-5 general rating
- 1-5 difficulty rating
- 1-100 Hours/week workload
- Review text (with Markdown support)

Additionally, I've added a simple verification mechanism to ensure that only GT students can submit reviews *without* requiring you to make an account.

## How Verification Works 🔒 

It's pretty straightforward. On the "Add Review" form, you'll see a new section labelled "Authentication".

<img width="1385" alt="Authentication section of 'Add Review' form" src="https://user-images.githubusercontent.com/19544466/183032565-82dbd059-db27-448f-b724-bf20df2c3de8.png">

Start typing your GT username and a "Send Code" button will pop up.

<img width="1385" alt="Authentication section of 'Add Review' form with username typed" src="https://user-images.githubusercontent.com/19544466/183032622-47fe4eb2-cd80-46f1-9045-599762e88d5a.png">

Click "Send Code" to have a six-digit code emailed to you at \<username\>@gatech.edu. Each code is valid for 10 minutes. When you receive the code, enter it into the field labelled "Code" and submit your review. That's it!

<img width="1385" alt="Authentication section of 'Add Review' form with username and code typed" src="https://user-images.githubusercontent.com/19544466/183032771-607abd74-2fb8-444f-9a51-3190946152be.png">

## Data Persistence & Stewardship 🔑 

Many folks rightfully expressed concerns over data stewardship during a previous iteration of this site. To clear the air, OMSCentral will never have a [paywall](https://www.reddit.com/r/OMSCS/comments/um1fd9/omscentralcom_is_down_ddos/). Your data is your data, and:
- I will not sell it to other students and not compensate you.
- You can delete your data at any time.

When you submit your GT username to my server to create a review - this is required to verify the one-time code - I encrypt it using [AES-256](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) with a key that is stored as an [environment variable on Vercel](https://vercel.com/docs/concepts/projects/environment-variables) (and is itself encrypted at rest by Vercel). Your encrypted username is stored on the backend CMS platform that I am using ([Sanity CMS](https://www.sanity.io/), for those interested) as the `authorID` for each review. This mechanism allows reviews are linked to GT students while keeping authors anonymous.

**IF YOU EVER WANT TO DELETE YOUR REVIEW**, please contact me privately at [matt@oms.fyi](mailto:matt@oms.fyi) from your GT account. I will re-encrypt your email using my key and, if the values match, I will delete your review. In the future, this mechanism may be more automated via the website. 

> Why encryption and not hashing? There may be a future in which folks have to option to toggle the anonymity of their reviews, and I would rather not obliterate the ability to recover a username from some derived value.

Unfortunately, I have *zero* identifying information for reviews created before I inherited this project, but I will make sure every new review is deletable via a quick verification process.

Thank you for using the site, and I hope to read your review soon.

On to the next feature,
\- Matt

**P.S.** For those interested in the implementation details, I wrote a quick [Tech Spec](https://www.dropbox.com/scl/fi/q3jlsyjusd5j2x860dhgk/Adding-Reviews.paper).

**Comment by MatthewCaseres** [comment]
grats, glad it is just the verified students that can leave reviews

**Reply by m4ttsch** [reply]
Thanks!
